### DISCUSSION TITLE
Need help compiling driver for Ceton InfiniTV 6

### DISCUSSION THREAD

**Discussion by Robert-Rathbone** [OpenSageTV/Community#33] [Tuners]
I need some help from any Linux experts reading this. I've had the InfiniTV 6 running on my SageTV server running Ubuntu Server 18.04 LTS (upgraded from earlier LTS versions) and it's been running fine for years. The boot drive on that machine is dying so I replaced the drive and installed Ubuntu Server 22.04.02 LTS.

I pulled the latest driver source code for the InfiniTV 6 from this Github repository: https://github.com/JamesRHarris/infinitv_pcie

I've tried compiling the driver and after getting all the prerequisites out of the way (installing gcc, make, linux headers) I was able to compile it with some warnings. I made a slight alteration to the code to fix the warnings, which hopefully fixed it without creating a bug (I cast a parameter to memcpy as a void pointer in ctn91xx_net.c). The resulting module seems to install ok but the driver doesn't seem to be working.

I discovered in /var/log/dmesg that the ctn0 is renamed to enp2s0. enp2s0 does not get assigned an IP address so I used netplan to set it to use DHCP, which gives it the IP address of 192.168.200.2.

I installed OpenDCT but it doesn't seem to be able to get any response from the InfiniTV 6 and SageTV of course doesn't see any video sources.

If you're willing to help me diagnose the problem I'm ready to answer any questions you have or run any diagnostic steps you suggest. My SageTV is dead in the water until this is fixed.

Thanks!
Robert Rathbone

**Comment by RJ-Make** [comment]
I know nothing about Linux, but have you tired the supplied [source form Ceton](http://cetoncorp.com/infinitv_support/linux_drivers/)?

**Reply by Robert-Rathbone** [reply]
Those drivers haven't been touched since 2013 and don't support the newer Linux kernel versions.  The GitHub repository for their source (https://github.com/ceton/infinitv_pcie) is a little more up-to-date, but it's still not had an update since 2018.  I've been using the source from someone who has forked the code and updated it as far as kernel version 5.18 (https://github.com/JamesRHarris/infinitv_pcie).  My kernel is at 5.19, so I don't know if there are any needed updates to make it compatible with 5.19.

My C programming skills are a bit rusty, so I don't know if the modifications I made helped or hurt.  It seems to compile and create an installable module, but I'm probably overlooking something obvious.

Thanks for the reply,
Robert

**Comment by Robert-Rathbone** [comment]
As a sort of a **bump** and to add new information on my problem:

I've been fiddling with this and using OpenDCT's "console-only" I watched its output and noticed that it's sending out uPnp broadcasts (I may not be getting the terminology right) on each network interface (the InfiniTV gets assigned 192.168.200.2) and waiting for a response that never comes.  Presumably the InfiniTV is supposed to respond by identifying itself and its capabilities but it never does.  I've been grasping as straws, making the the firewall is disabled in case it was preventing packets from getting sent to the InfiniTV.  I think the firewall only blocks network traffic to/from other machines and not internal, but I could be wrong.  I can successfully ping the InfiniTVs address but as best as I can tell it's not responding in any other way.

I'm getting desperate because I'm pushing the limits of what I know about Linux's inner workings.  I don't know if the problem is in the compiling of the driver and/or its compatibility with the 5.19 kernel or if there's something in the OS that needs to be configured differently, installed or uninstalled in order for this to start working.

Thank you very much if you can offer any help!

Robert

**Comment by emveepee** [comment]
I only understand this via the support provided NextPVR users who want to use opendct with NextPVR but from the opendct logs that I have seen the six tuners were on one IP.

```
16:33:47.114 [cling-20] INFO  InfiniTVCaptureDevice - Encoder Manufacturer: 'Ceton', Number: 1, Remote IP: '192.168.2.250', Local IP: '192.168.2.253', CableCARD: true, Lineup: 'dct_infinitv', Offline Scan Enabled: false, RTP Port: 8300
16:33:47.174 [cling-20] INFO  InfiniTVCaptureDevice - Encoder Manufacturer: 'Ceton', Number: 2, Remote IP: '192.168.2.250', Local IP: '192.168.2.253', CableCARD: true, Lineup: 'dct_infinitv', Offline Scan Enabled: false, RTP Port: 8488
16:33:47.221 [cling-20] INFO  InfiniTVCaptureDevice - Encoder Manufacturer: 'Ceton', Number: 3, Remote IP: '192.168.2.250', Local IP: '192.168.2.253', CableCARD: true, Lineup: 'dct_infinitv', Offline Scan Enabled: false, RTP Port: 8344
16:33:47.267 [cling-20] INFO  InfiniTVCaptureDevice - Encoder Manufacturer: 'Ceton', Number: 4, Remote IP: '192.168.2.250', Local IP: '192.168.2.253', CableCARD: true, Lineup: 'dct_infinitv', Offline Scan Enabled: false, RTP Port: 8302
16:33:47.316 [cling-20] INFO  InfiniTVCaptureDevice - Encoder Manufacturer: 'Ceton', Number: 5, Remote IP: '192.168.2.250', Local IP: '192.168.2.253', CableCARD: true, Lineup: 'dct_infinitv', Offline Scan Enabled: false, RTP Port: 8456
16:33:47.392 [cling-20] INFO  InfiniTVCaptureDevice - Encoder Manufacturer: 'Ceton', Number: 6, Remote IP: '192.168.2.250', Local IP: '192.168.2.253', CableCARD: true, Lineup: 'dct_infinitv', Offline Scan Enabled: false, RTP Port: 8386

```

**Reply by Robert-Rathbone** [reply]
Thanks for the reply.  I was aware that the tuner uses a single IP address but your comment gave me the idea that perhaps the appropriate ports were close or being blocked by the OS.  I can ping the address of the tuner and it responds but doing an nmap of the IP address shows a lot of ports being closed.  I don't know what ports this tuner needs open so I can't examine those specifically.  I don't know if the ports are closed because the OS is blocking them or if they are closed because the tuner isn't using them (perhaps because the driver didn't compile correctly).  I played with using the "ufw" command to enable and try opening ports but it doesn't seem to change anything so far so I gave up on that.

You've also given me the idea of checking into NextPVR and see what experience anyone there has with the InfiniTV tuner.

Thanks,
Robert

**Comment by emveepee** [comment]
Sorry I thought your issues was with multiple IP, with OpenDCT it should work, firewalls aside  I wouldn't bother with NextPVR since you still would need a working OpenDCT.  Internal DCT support was dropped years ok.  Ceton Proxy might be another option.

Martin

**Comment by Robert-Rathbone** [comment]
No problem.  I think my problem is before OpenDCT comes into play...  It just doesn't seem to be responding to the uPnP request from OpenDCT so that it can be recognized as a tuner.  The only thing I've found that it responds to are pings.  I tried connecting to the web page it's supposed to host on it's IP address and get nothing.  I'm doing everything from the command line (it's a server OS, so no GUI) and so I have to do everything from the terminal.  Thus I can only use part of the test they give in the driver's README file:
cat /dev/ceton/ctn91xx_mpeg0_0 | mplayer -cache 8192 -

I tried the "cat /dev/ceton/ctn91xx_mpeg0_0" command, expecting to see data scrolling down the screen, but get nothing.  So I believe that either the driver is at fault or the OS is preventing it from working.  Hopefully someone with some experience in these things will see this and have some ideas...

Thanks,
Robert

**Reply by Robert-Rathbone** [reply]
I just made a discovery that might help...  My InfiniTV is set to 192.168.200.2 and if I issue a "ifconfig -a" command, it shows that ctn0 has that IP address.  However, I decided to try to do a uPnP search myself and issued the following command (I had to Google for the command):

`gssdp-discover -i ctn0 --timeout=3`

Interestingly this command discovered that the InfiniTV was responding on 192.168.200.1, and I was able to use lynx (the command line web browser since I have no GUI) to open the tuner's web page.

Oddly, OpenDCT does not see this IP address to send a uPnP discovery request to that address and I don't know how to get it to see it.  Even "ifconfig -a" does not show it, but the "gssdp-discover" command issued against the "ctn0" interface does see it.

I've been hoping the author of OpenDCT, EnterNoEscape, might see this and offer some suggestions.

Thanks,
Robert
